### ISSUE TITLE
new feature wanted; Detect new version.

### ISSUE THREAD

**Issue by zander** [bitcoinclassic/bitcoinclassic#221]
For users of bitcoin-qt (The GUI version) I think we can have a simple way to detect a new version.

* on first start, ask the user if they want to check online for a new version of Bitcoin Classic.
if they user says yes, write this in the config file.

* on every startup of Bitcoin-qt (but no more than once a day), check that config file on the user having asked for a version-check and if so download a simple text file from bitcoinclassic.com which contains the version number and maybe the date of the latest released version.
If this is newer than we are running, notify the user of this fact and suggest they upgrade. This is only a dialog. No auto-downloads or install should happen. The user should be able to open a browser that goes to the download page directly. (https://bitcoinclassic.com/downloads)

**Comment by ghost** [comment]
I think we should be wary of writing data to a configuration file as this is mixing static data with dynamic data.
You could argue that the new version online check is just a once write configuration parameter, but how would you know when you last performed a version check? This needs to be stored somewhere.
It would probably be better to write this information into a new dynamic storage file.
The reason for this, is if I setup a node, I will probably copy an existing configuration file across. So we don't want to set a bad precedent by having any data if the configuration file that is frequently altered on the fly.

**Comment by ghost** [comment]
Perhaps I misunderstood the configuration file you was referring too? On linux there is the ~/.config/Bitcoin-Qt.conf file which could be used to store information like this, and may be overriden by a setting in [datadir]/bitcoin.conf.

**Comment by zander** [comment]
I indeed meant the bitcoin-qt.conf file as that one is generated by the Qt client, not the user.

**Comment by ghost** [comment]
Presumably this is written through a QT library function, as on Windows I think these settings are stored in the registry.
Perhaps a separate issue, but these settings are shared between multiple QT nodes that are run on the same machine. Would love to see most of these settings moved to the data directory, but at least some of them can be overridden in the bitcoin.conf file stored in the data directory.

**Comment by zander** [comment]
@CryptoDwarf Lets keep things on topic, GitHub is a really bad platform to have open conversations.
You may want find a more real-time platform for talks; see the bottom parags of https://bitcoinclassic.com/community.html for suggestions.

**Comment by gubatron** [comment]
I have some experience i this subject.

From a UI perspective I wouldnt show that question on the first run, eventually you will want to do the same with other settings and it will become a nightmare to first run the client at a tome where it will be downloading the blockchain.

Make it an opt-out setting (automatically checking for updates on startup) and make the update process dismissable. When the update is available you can have a subtle icon letting you know an update is available and the user can chose to update or not (as automatic updates are always a huge vulnerability to computer networks.

Second, if you are willing to make the version check depending on bitcoinclassic.com, make sure it reads a gpg signed metadata file, say bitcoinclassic.com/version.json (that would also contain  information like a changelog, the urgency of the update, and different build numbers for different OS distributions).

however I'd stay away from this centralized approach and I'd make the spread of update messages via p2p signed viral messages that could be spread as clients do their handshakes. If your client encounters another bitcoinclassic peer with a seeming higher build/version number it could ask such peer for its signed update message. If the message is valid, then the update notification icon is shown in the UI.

The distribution of the binary should be another subject

**Comment by NiKiZe** [comment]
Any automatic updates should be opt-in IMHO and not opt-out.
I would also suggest that this has a compile time opt-in option, and without it no code at all should be included in the resulting binary. (some security centric users will probably refuse to run any client that has any kind of "phone home" feature even included, regardless of how good the intentions are)

**Comment by zander** [comment]
I want to acknowledge the idea that "opt-out" is going to frighten a certain set of users as long as there is a centralized server they connect to. A so called "calling home" label is something I'd like to avoid.

I'm interested in the idea of doing it distributed, which would avoid the issue. It reminds me of the event broadcast system we had for many years in Bitcoin. But we retired because it was in the hands of a few wrong men. It might be Ok to come up with a replacement system.

It will likely be much easier to sign using a bitcoin-address/private key than GPG.

**Comment by gubatron** [comment]
>Any automatic updates should be opt-in IMHO and not opt-out.

I agree 100%, however, let's make clear the difference between letting the user know there is an update, vs downloading and update automatically, vs applying an automatic update.

I believe the notification should be shown by default (opt-out, "don't show me this again" type of thing), and the notification should be a subtle icon on the status bar, and if we have a notification system that differentiates depending on the urgency of the update the notification mechanism could be a bit more intrusive, say there's a huge vulnerability bug with your current version, perhaps on the next restart you should get a dialog letting you know you're at risk and why you should update, the update-deceision is still up to the user.

**Comment by NiKiZe** [comment]
@gubatron no not even showing an update is ok, the reason for this is that the server that holds information about an update will get a "phone home" from each client that is running and that is a potential information leak that would be unacceptable.

Also if I compile from source, I certainly don't even want to have the option to run a "unknown binary" downloaded from someone else, For example I could have local modifications. (thus this feature most be opt-in at compile time)

I was thinking a bit, and maybe this could be done via p2p, instead of the alert system that once existed I wonder if user-agent could be used to identify by other clients that a new version is out there in the wild. and downloading that update could also (in theory) be done by fetching itself from other clients.
Let's say that "I see x other nodes with a newer version for matching user-agent" that could be used to show that notification without any "phone home". The downside is that propagation is slow, and it is also possible to Sybil or interrupt nodes that use this for version detection.

**Comment by gubatron** [comment]
>@gubatron no not even showing an update is ok, the reason for this is that the server that holds information about an update will get a "phone home" from each client that is running and that is a potential information leak that would be unacceptable

I see, very valid concern. Then we should simply take off the table any update message fetching from centralized web servers.

Update messages should be sent as viral messages that spread throughout the network as clients perform handshakes and learn that their last known message is old, in a similar way to what you mention.

You shouldn't necessarily wait to see a newer client to be aware of the latest update message, since many of the clients might not want to update right away and there's a chance you won't ever connect to a fresh new client, like you said, this would make propagation pretty slow :)

I wouldn't be concerned about Sybil attacks since the update message should always be a signed message.

However, all clients should quickly learn what the latest update message is.

There's pretty easy ways of doing this with the libtorrent DHT.

The libtorrent DHT has both immutable and mutable entries. Mutable entries are only modifiable by whoever has access to the entry's key. If the latest update message is stored in a key that every client knows and trusts (the content of the key is signed, etc.) all clients can perform a DHT query for the latest contents and know if their locally stored version of the message is old.

I've other ideas on how libtorrent could be used to improve block distribution but that falls out of this topic.

**Comment by zander** [comment]
Some thoughts;

* this feature in total is only useful to a subset of clients. People that install from their distro directly don't care as they get an update notification elsewhere. Similarly, people that use bitcoind instead of the GUI client probably won't care either. Agreed with above about self-compiled clients. This needs a opt-in config flag.

* I highly recommend starting simple and building this out according to need.

* A centralized solution is not perfect, and not for everyone. But! The people that have requested this solution did, however, state they would be happy with a centralized solution. So if its opt-in, I think it may be a good first step. Because its simple.

* The bitcoinclassic.com website is protected by cloudflare, this may interfere with simple requests from popular networks (try over tor to see the captcha pop up).

* This task is about showing a changelog and a link to the website. Again, focus on the simple first.

* A way to replace the current running version automatically is something that needs plenty of security expertise to do securely. Downloading and installing software requires privilege elevation and different strategy and code on each major OS.

**Comment by gubatron** [comment]
>this feature in total is only useful to a subset of clients. 

agreed, it could only be included if configured at compilation time, and not included in default builds.

**Comment by DeftNerd** [comment]
If I could make a suggestion, instead of checking the Bitcoin Classic website (which could have Cloudflare problems especially if the client is behind Tor), perhaps you could use the Bitcoin network to indicate a new version somehow. 

For example, if there was a "Bitcoin Classic" donation address hard coded into the client, perhaps the client could monitor the address for a transaction to the address from another trusted address (yours?) and use the amount transacted to determine the version number. 

A transaction of 135000 satoshi from the trusted address to the donation address would indicate that version 1.3.5 is now available and some kind of notification could be logged or placed in the client interface. 

It would allow for version signaling without accessing any different network resources.

**Comment by zander** [comment]
@DeftNerd I think thats genius, definitely is something to look into.
